VisualPinball.Unity/VisualPinball.Unity.Editor/Editors/DragPointsEditor.cs

```diff
@@ -1,522 +1,650 @@
+				}
+
+				return true;
+			}
+
+			[MenuItem(CONTROLPOINTS_MENUPATH + "/IsSmooth", false, 1)]
+			private static void Smooth(MenuCommand command)
+			{
+				ItemInspector editor = command.context as ItemInspector;
+				if (editor == null || editor.DragPointsEditor == null){
+					return;
+				}
+
+				var dpoint = RetrieveDragPoint(editor.DragPointsEditor, command.userData);
+				if (dpoint != null){
+					editor.DragPointsEditor.PrepareUndo("Changing DragPoint IsSmooth");
+					dpoint.IsSmooth = !dpoint.IsSmooth;
+				}
+			}
+
+			[MenuItem(CONTROLPOINTS_MENUPATH + "/IsSmooth", true)]
+			private static bool SmoothValidate(MenuCommand command)
+			{
+				ItemInspector editor = command.context as ItemInspector;
+				if (editor == null || editor.DragPointsEditor == null || editor.DragPointsEditor.IsItemLocked())
+				{
+					return false;
+				}
+
+				if (!editor.DragPointsEditor.HasDragPointExposition(DragPointExposition.Smooth))
+				{
+					Menu.SetChecked($"{CONTROLPOINTS_MENUPATH}/IsSmooth", false);
+					return false;
+				}
+
+				var dpoint = RetrieveDragPoint(editor.DragPointsEditor, command.userData);
+				if (dpoint != null){
+					Menu.SetChecked($"{CONTROLPOINTS_MENUPATH}/IsSmooth", dpoint.IsSmooth);
+				}
+
+				return true;
+			}
+
+			[MenuItem(CONTROLPOINTS_MENUPATH + "/Remove Point", false, 101)]
+			private static void RemoveDP(MenuCommand command)
+			{
+				ItemInspector editor = command.context as ItemInspector;
+				if (editor == null || editor.DragPointsEditor == null)
+				{
+					return;
+				}
+
+				if (EditorUtility.DisplayDialog("DragPoint Removal", "Are you sure you want to remove this Dragpoint ?", "Yes", "No"))
+				{
+					editor.DragPointsEditor.RemoveDragPoint(command.userData);
+				}
+			}
+
+			[MenuItem(CONTROLPOINTS_MENUPATH + "/Remove Point", true)]
+			private static bool RemoveDPValidate(MenuCommand command)
+			{
+				ItemInspector editor = command.context as ItemInspector;
+				if (editor == null || editor.DragPointsEditor == null || editor.DragPointsEditor.IsItemLocked())
+				{
+					return false;
+				}
+
+				return true;
+			}
+
+			//Curve Traveller
+			[MenuItem(CURVETRAVELLER_MENUPATH + "/Add Point")]
+			static void AddDP(MenuCommand command)
+			{
+				ItemInspector editor = command.context as ItemInspector;
+				if (editor == null || editor.DragPointsEditor == null)
+				{
+					return;
+				}
+
+				editor.DragPointsEditor.AddDragPointOnTraveller();
+			}
+		}
+
+		public bool IsItemLocked()
+		{
+			IEditableItemBehavior editable = _target as IEditableItemBehavior;
+			if (editable == null)
+			{
+				return true;
+			}
+			return editable.IsLocked;
+		}
+
+		public bool HasDragPointExposition(DragPointExposition dpExpo)
+		{
+			IDragPointsEditable dpeditable = _target as IDragPointsEditable;
+			if (dpeditable == null)
+			{
+				return false;
+			}
+			return (dpeditable.GetDragPointExposition() & dpExpo) != DragPointExposition.None;
```

Comment:
This is... tricky. What's the logic behind `DragPointExposition`? I wouldn't expect boolean operations on enums here. Can this be simplified?

Reply:
Each Items using drag points don't expose the same data from them.
Ramps : Smooth, SlingShot
Surfaces : Smooth Slingshot, Texturing data
Rubbers : Smooth
Triggers : Smooth, Slingshot
Because of these combinations i used flags to expose features from the DragPointsEditor instance linked to the Item type.
It could also be some IsUsingSmooth, IsUsingSlingShot & IsUsingTexture calls if you prefer or even something else.

Reply:
Or just a set with the enums, and use [`Contains`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1.contains?view=netframework-4.7.2)? I didn't even know that ramps can do slingshots as well...

Reply:
> Or just a set with the enums, and use [`Contains`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1.contains?view=netframework-4.7.2)? I didn't even know that ramps can do slingshots as well...

Yup will change that.
About ramps i don't know, i just went back to VPX and tried any items to see what's exposed for drag points. Maybe it's already a bug or was never used.
Also, when rendering the Slingshot segments in red on curve display, without taking into account the slingshot property exposure, i saw slingshot segments in Rubbers which should only expose Smooth. 
Or maybe i didn't check te correct way in VPX

Reply:
Hmm okay, weird. I'll have a look the VPX source to check whether slingshots are used otherwise than on surfaces. Ramps (the non-wire ones) are actually used for all kind of stuff in VPX tables, so it wouldn't surprise me if they supported slingshots.

Reply:
Yeah actually they are. They will just generate a slingshot hit object, which acts the same as if generated through a surface.